src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -984,6 +984,253 @@ private Task WriteSuffix()
+            {
+                if (limitedBuffer.Length == ServerOptions.Limits.MaxRequestLineSize)
+                {
+                    RejectRequest(RequestRejectionReason.RequestLineTooLong);
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            const int stackAllocLimit = 512;
+
+            // Move 1 byte past the \r
+            end = limitedBuffer.Move(end, 1);
+            var startLineBuffer = limitedBuffer.Slice(0, end);
+
+            Span<byte> span;
+
+            if (startLineBuffer.IsSingleSpan)
+            {
+                // No copies, directly use the one and only span
+                span = startLineBuffer.ToSpan();
+            }
+            else if (startLineBuffer.Length < stackAllocLimit)
+            {
+                unsafe
+                {
+                    // Multiple buffers and < stackAllocLimit, copy into a stack buffer
+                    byte* stackBuffer = stackalloc byte[startLineBuffer.Length];
+                    span = new Span<byte>(stackBuffer, startLineBuffer.Length);
+                    startLineBuffer.CopyTo(span);
+                }
+            }
+            else
+            {
+                // We're not a single span here but we can use pooled arrays to avoid allocations in the rare case
+                span = new Span<byte>(new byte[startLineBuffer.Length]);
+                startLineBuffer.CopyTo(span);
+            }
+
+            var methodEnd = 0;
+            if (!span.GetKnownMethod(out string method))
+            {
+                methodEnd = span.IndexOf(ByteSpace);
+                if (methodEnd == -1)
+                {
+                    RejectRequestLine(start, end);
+                }
+
+                method = span.Slice(0, methodEnd).GetAsciiString();
+
+                if (method == null)
+                {
+                    RejectRequestLine(start, end);
+                }
+
+                // Note: We're not in the fast path any more (GetKnownMethod should have handled any HTTP Method we're aware of)
+                // So we can be a tiny bit slower and more careful here.
+                for (int i = 0; i < method.Length; i++)
+                {
+                    if (!IsValidTokenChar(method[i]))
+                    {
+                        RejectRequestLine(start, end);
+                    }
+                }
+            }
+            else
+            {
+                methodEnd += method.Length;
+            }
+
+            var needDecode = false;
+            var pathBegin = methodEnd + 1;
+            var pathToEndSpan = span.Slice(pathBegin, span.Length - pathBegin);
+            // var pbs = Encoding.UTF8.GetString(pathToEndSpan.ToArray());
+            pathBegin = 0;
+
+            // TODO: IndexOfAny
+            var spaceIndex = pathToEndSpan.IndexOf(ByteSpace);
```

Comment:
Split current `IndexOf` methods in two parts, one that can search byte (or any of 3 bytes ) in span, and another that calls it on all spans inside `ReadableBuffer` and returns `ReadCursor`